File: Napakalaki_Java/src/napakalaki/Napakalaki.java
/*
********************************************************************
**       _   _                   _         _       _    _         **
**      | \ | | __ _ _ __   __ _| | ____ _| | __ _| | _(_)        **
**      |  \| |/ _` | '_ \ / _` | |/ / _` | |/ _` | |/ / |        **
**      | |\  | (_| | |_) | (_| |   < (_| | | (_| |   <| |        **
**      |_| \_|\__,_| .__/ \__,_|_|\_\__,_|_|\__,_|_|\_\_|        **
**                  |_|                                           **
** Creado por Javier Bolívar Valverde <javibolivar@correo.ugr.es> **
** Programación y Diseño Orientado a Objetos                      **
** Grado en Ingeniería Informática (Universidad de Granada)       **
********************************************************************
*/
package napakalaki;

import java.util.ArrayList;
import java.util.Random;


public class Napakalaki {
    //Atributos
    private static final Napakalaki instance = null;
    private Monster currentMonster;
    private CardDealer dealer;
    private ArrayList<Player> players;
    private Player currentPlayer;
    
    //Métodos
    private Napakalaki(){
    }
    
    public static Napakalaki getInstance() {
        return instance;
    }
    
    private void initPlayers(ArrayList<String> names) {
        this.players = new ArrayList();
        for(String name : names) {
            Player playeract = new Player(name);
            this.players.add(playeract);
        }
        nextPlayer(); //Para rellenar currentPlayer
    }
    
    private Player nextPlayer() {
        int tamplayers = this.players.size();
        int indice; //Índice del siguiente jugador.
        //Si no hay un jugador actual, es la primera jugada.
        if (this.currentPlayer == null) {
            Random random = new Random();
            //Aleatorio = número entre 0 y (tamplayers - 1)
            indice = random.nextInt(tamplayers);
        }
        else {
            indice = this.players.indexOf(this.currentPlayer);
            //Aplicamos el índice módulo tamplayers para que se limite al array de jugadores que tenemos.
            indice = (indice + 1) % tamplayers;
        }
        
        return this.players.get(indice);
    }
    
    private boolean nextTurnAllowed() {
        boolean allow = false;
        
        //Si es la primera jugada, también se permite el turno.
        if (this.currentPlayer == null || this.currentPlayer.validState())
            allow = true;
        return allow;
    }
    
    private void setEnemies(){
        int tamplayers = this.players.size();
        Random random = new Random();
        Player enemy, playeract;
        //Recorremos todo el ArrayList
        for (Player player : this.players) {
            int i = this.players.indexOf(player);
            int j;
            do {
                //j= número aleatorio entre 0 y tamaño de players(excluido)
                j = random.nextInt(this.players.size());
            }while(j==i); //Comprueba que el jugador actual no sea el mismo que su enemigo.
            
            this.players.get(i).setEnemy(this.players.get(j));
        }
    }
    
    public CombatResult developCombat() {
        Monster m = this.currentMonster;
        CombatResult combatResult = this.currentPlayer.combat(m);
        this.dealer.giveMonsterBack(m);
        
        
        return combatResult;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures) {
        for(Treasure treasure : treasures) {
            this.currentPlayer.discardVisibleTreasure(treasure);
            this.dealer.giveTreasureBack(treasure);
        }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures) {
        for(Treasure treasure : treasures) {
            this.currentPlayer.discardHiddenTreasure(treasure);
            this.dealer.giveTreasureBack(treasure);
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures) {
        
    }
    
    public void initGame(ArrayList<String> players) {
        this.initPlayers(players);
        this.setEnemies();
        this.dealer.initCards();
        this.nextTurn();
    }
    
    public Player getCurrentPlayer() {
        return this.currentPlayer;
    }
    
    public Monster getCurrentMonster() {
        return this.currentMonster;
    }
    
    public boolean nextTurn() {
        boolean stateOK = this.nextTurnAllowed();
        if(stateOK) {
            this.currentMonster = this.dealer.nextMonster();
            this.currentPlayer = this.nextPlayer();
            boolean dead = this.currentPlayer.isDead();
            if(dead){
                this.currentPlayer.initTreasures();
            }
        }
        
        return stateOK;
    }
    
    //Compara result con el valor WINGAME del enumerado
    //Devuelve true si result == WINGAME, false en caso contrario
    public boolean endOfGame(CombatResult result) {
        return result == CombatResult.WINGAME;
    }
}

File: Napakalaki_Ruby/lib/napakalaki.rb
# encoding: utf-8

=begin
********************************************************************
**       _   _                   _         _       _    _         **
**      | \ | | __ _ _ __   __ _| | ____ _| | __ _| | _(_)        **
**      |  \| |/ _` | '_ \ / _` | |/ / _` | |/ _` | |/ / |        **
**      | |\  | (_| | |_) | (_| |   < (_| | | (_| |   <| |        **
**      |_| \_|\__,_| .__/ \__,_|_|\_\__,_|_|\__,_|_|\_\_|        **
**                  |_|                                           **
** Creado por Javier Bolívar Valverde <javibolivar@correo.ugr.es> **
** Programación y Diseño Orientado a Objetos                      **
** Grado en Ingeniería Informática (Universidad de Granada)       **
********************************************************************
=end

require 'singleton'
require_relative 'player'
require_relative 'card_dealer'

class Napakalaki
  include Singleton

  attr_accessor :current_player, :players, :current_monster, :dealer

  private
  def init_players(names)
    @dealer = CardDealer.instance
    @players = Array.new
    names.each do |name|
      @players << Player.new(name)
    end
    next_player
  end

  def next_player
    tamplayers = @players.size

    if (@current_player == nil)
      indice = rand(tamplayers)
    else
      indice = @players.index(@current_player)

      indice = (indice + 1) % tamplayers
    end
    return @players.at(indice)
  end

  def next_turn_allowed
    allow = false
    allow = true if (@current_player == nil || @current_player.valid_state)
    return allow
  end

  def set_enemies
    tamplayers = @players.size;
    players.each do |player|
      i = @players.index(player)
      j = rand()
      loop do
        j = rand(@players.size)
        break if (j==i)
      end

      @players[i].enemy = @players[j]
    end
  end

  public
  def develop_combat
    m = @current_monster
    combat_result = @current_player.combat(m)
    @dealer.give_monster_back(m)

    combat_result
  end

  def discard_visible_treasures(treasures)
    treasures.each do |treasure|
      @current_player.discard_visible_treasure(treasure)
      @dealer.give_treasure_back(treasure)
    end
  end

  def discard_hidden_treasures(treasures)
    treasures.each do |treasure|
      @current_player.discard_hidden_treasure(treasure)
      @dealer.give_treasure_back(treasure)
    end
  end

  def make_treasures_visible(treasures)
    treasures.each do |t|
      @current_player.make_treasure_visible(t)
    end
  end

  def init_game(players)
    init_players(players)
    set_enemies
    @dealer.init_cards
    next_turn
  end

  def next_turn
    state_ok = next_turn_allowed
    if(state_ok)
      @current_monster = @dealer.next_monster
      @current_player = next_player
      dead = @current_player.dead
      if(dead)
        @current_player.init_treasures
      end
    end

    state_ok
  end

  def end_of_game(result)
    return result == WINGAME
  end
end
